packages/test/src/workflows/signals-are-always-processed.ts

```diff
@@ -0,0 +1,19 @@
+import { proxySinks, Sinks, setHandler, defineSignal } from '@temporalio/workflow';
+
+export const incrementSignal = defineSignal('increment');
+
+export interface SignalProcessTestSinks extends Sinks {
+  controller: {
+    sendSignal(): void;
+  };
+}
+
+const { controller } = proxySinks<SignalProcessTestSinks>();
+
+export async function signalsAreAlwaysProcessed(): Promise<void> {
+  let counter = 0;
+  setHandler(incrementSignal, () => void ++counter);
+  if (counter === 0) {
+    controller.sendSignal();
```

Comment:
Pardon my lack of TS SDK knowledge wrt sinks. So this asynchronously invokes the `sendSignal` call in the worker JS? And of course since this is deterministic, so it surely doesn't wait for some non-deterministic worker-side stuff right? So if it doesn't wait on anything, this simply is like a beacon to the worker-side and then this workflow exits immediately?

Reply:
The sink function is invoked and blocks workflow task completion.
If a signal is received while workflow task is being processed the task will fail and the workflow will need to be replayed.

Reply:
Hrmm, this can definitely be confusing for normal JS writers. A synchronous call that magically blocks for an internal asynchronous call seems off from all other JS.

Also, is it ok that these sinks may do something non-deterministic? What if it errors some of the time? Or is this like side effects in other SDKs where this is only executed once and result is stored in history?

Reply:
The synchronous call doesn't block, it's buffered in the workflow runtime.
When the worker completes the activation and the isolate finishes its work, the worker will pull out all of the buffered sink calls and run the injected sink functions.
The functions will block the worker from completing the workflow.
Sinks cannot return values back to the workflow or generate any workflow commands.

Reply:
If a sink function throws we log it and ignore the error.

Reply:
> The synchronous call doesn't block, it's buffered in the workflow runtime.

To confirm, from the perspective of the JS writer here, that `controller.sendSignal()` does not return until the sink call is done? Since this blocks in a single-threaded runtime, how can I do something like `Promise.race([sink.sendSignal1(), sink.sendSignal2()])`? Or is it known that sink will pause the entire runtime, regardless of whether the sink call itself may be doing something asynchronous, and therefore two sink calls cannot run at once, even if they do something asynchronous like network IO?

Reply:
Let’s set up some time and I’ll brief you about how this works and why it’s done that way.
Same as for why signals are handled via callbacks